text/0000-hash-simplification.md

```diff
@@ -0,0 +1,420 @@
+defined for `Hash<H> where H: Writer + Hasher`, essentially forcing a
+byte-stream oriented hashing algorithm for all hashing.
+
+Some examples of using this API are:
+
+```rust
+use std::hash::{Hash, Hasher, Writer, SipHasher};
+
+impl<S: Hasher + Writer> Hash<S> for MyType {
+    fn hash(&self, s: &mut S) {
+        self.field1.hash(s);
+        // don't want to hash field2
+        self.field3.hash(s);
+    }
+}
+
+fn sip_hash<T: Hash<SipHasher>>(t: &T) -> u64 {
+    let mut s = SipHasher::new_with_keys(0, 0);
+    t.hash(&mut s);
+    s.finish()
+}
+```
+
+Forcing many `impl` blocks to require `Hasher + Writer` becomes onerous over
+times and also requires at least 3 imports for a custom implementation of
+`hash`. Taking a generically hashable `T` is also somewhat cumbersome,
+especially if the hashing algorithm isn't known in advance.
+
+Overall the `std::hash` API is generic enough that its usage is somewhat verbose
+and becomes tiresome over time to work with. This RFC strives to make this API
+easier to work with.
+
+## Forcing byte-stream oriented hashing
+
+Much of the `std::hash` API today is oriented around hashing a stream of bytes
+(blocks of `&[u8]`). This is not a hard requirement by the API (discussed
+above), but in practice this is essentially what happens everywhere. This form
+of hashing is not always the most efficient, although it is often one of the
+more flexible forms of hashing.
+
+Other languages such as Java and C++ have a hashing API that looks more like:
+
+```rust
+trait Hash {
+    fn hash(&self) -> usize;
+}
+```
+
+This expression of hashing is not byte-oriented but is also much less generic
+(an algorithm for hashing is predetermined by the type itself). This API is
+encodable with today's traits as:
+
+```rust
+struct Slot(u64);
+
+impl Hash<Slot> for MyType {
+    fn hash(&self, slot: &mut Slot) {
+        *slot = Slot(self.precomputed_hash);
+    }
+}
+
+impl Hasher for Slot {
+    type Output = u64;
+    fn reset(&mut self) { *self = Slot(0); }
+    fn finish(&self) -> u64 { self.0 }
+}
+```
+
+This form of hashing (which is useful for performance sometimes) is difficult to
+work with primarily because of the frequent bounds on `Writer` for hashing.
+
+## Non-applicability for well-known hashing algorithms
+
+One of the current aspirations for the `std::hash` module was to be appropriate
+for hashing algorithms such as MD5, SHA\*, etc. The current API has proven
+inadequate, however, for the primary reason of hashing being so generic. For
+example it should in theory be possible to calculate the SHA1 hash of a byte
+slice via:
+
+```rust
+let data: &[u8] = ...;
+let hash = std::hash::hash::<&[u8], Sha1>(data);
+```
+
+There are a number of pitfalls to this approach:
+
+* Due to slices being able to be hashed generically, each byte will be written
+  individually to the `Sha1` state, which is likely to not be very efficient.
+* Due to slices being able to be hashed generically, the length of the slice is
```

Comment:
There's an alternate approach to this, similar to how Haskell makes its strings print as strings despite just being normal lists of characters.

```rust
trait Hash {
     fn hash<H: Hasher>(&self, h: &mut H);
     fn hash_slice<H: Hasher>(self_: &[Self], h: &mut H) {
          for x in self_ { x.hash(h) }
     }
}

impl Hash for u8 {
     fn hash<H: Hasher>(&self, h: &mut H) { h.write_u8(*self) } 
 
     fn hash_slice<H: Hasher>(self_: &[u8], h: &mut H) { h.write(self_) }
}

impl<'a, T: Hash> Hash for &'a [T] {
     fn hash<H: Hasher>(&self, h: &mut H) {
          Hash::hash_slice(*self, h);
     }
}
```

This allows any sequence of contiguous things to ensure they're hashed quickly.

Downside: a custom implementation of `hash_slice` may change behaviour, for some hashing algorithms.